packages/flutter/lib/src/widgets/inherited_model.dart

```diff
@@ -0,0 +1,212 @@
+  /// of the [dependencies].
+  @protected
+  bool updateShouldNotifyDependent(
+    covariant InheritedModel<T> oldWidget, Set<T> dependencies);
+
+  // Return the first ancestor of context whose widget is type T and for which
+  // visitor(widget) returns false. If visitor is null then this function is the
+  // same as: context.ancestorInheritedElementForWidgetOfExactType(T).
+  static InheritedElement _findModel<T extends InheritedModel<Object>>(
+    BuildContext context,
+    bool visitor(T widget)
+  ) {
+    final InheritedElement model = context.ancestorInheritedElementForWidgetOfExactType(T);
+    if (model == null)
+      return null;
+
+    assert(model.widget is T);
+    final T modelWidget = model.widget;
+    if (visitor == null || !visitor(modelWidget))
+      return model;
+
+    Element modelParent;
+    model.visitAncestorElements((Element ancestor) {
+      modelParent = ancestor;
+      return false;
+    });
+    if (modelParent == null)
+      return null;
+    return _findModel<T>(modelParent, visitor);
+  }
+
+  /// Makes [context] dependent on the specified [aspect] of an [InheritedModel]
+  /// of type T.
+  ///
+  /// When the given [aspect] of the model changes, the context will be
+  /// rebuilt. The [updateShouldNotifyDependent] method must determine if a
+  /// change in the model widget corresponds to an `aspect` value.
+  ///
+  /// By default, the dependency created by this method targets the first
+  /// [InheritedModel] ancestor of type T. If [visitor] is specified then
+  /// it's the first [InheritedModel] ancestor of type T for which
+  /// [vistor] returns false. The vistor parameter can be used to create
+  /// models that override some values and delegate the rest to an
+  /// ancestor model.
+  ///
+  /// In the following example the model's static `valueOf` method returns the
+  /// first non-null value of `path`. By doing so the caller will have created
+  /// a `context` dependency on the first model ancestor that has a non-null
+  /// value for the specified path.
+  ///
+  /// ```dart
+  /// class IntegerModel extends InheritedModel<String> {
+  ///   // ...
+  ///   int valueOf(String path) => _model.valueOf(path);
+  ///
+  ///   static int valueOf(BuildContext context, String path) {
+  ///     int value;
+  ///     InheritedModel.inheritFrom<IntegerModel>(context,
+  ///       aspect: path,
+  ///       visitor: (IntegerModel widget) => (value = widget.valueOf(path)) == null,
+  ///     );
+  ///     return value;
+  ///   }
+  /// }
+  /// ```
+  ///
+  /// If [aspect] and [visitor] are null this method is the same as:
+  /// `context.inheritFromWidgetOfExactType(T)`.
+  static T inheritFrom<T extends InheritedModel<Object>>(
+    BuildContext context, {
+    Object aspect,
+    bool visitor(T widget),
+  }) {
+    final InheritedElement model = _findModel<T>(context, visitor);
+    if (model != null)
+      return context.inheritFromWidgetOfExactType(T, aspect: aspect, target: model);
```

Comment:
one problem with this approach is that if one of the widgets in the middle is updated to now have the aspect you're looking for, you're not going to get updated.

i.e. consider a tree A->B->C->D->E where A and C and inherited models M, only A provides aspect X, and E depends on model M aspect X. Now update the tree so that only B and C update, and C now has model X; E won't be notified and won't rebuild.

Reply:
That's true. Dynamically changing the list of aspects a model supports would require rebuilding its subtree. It's comparable to changing the tree by adding or removing an inherited widget.

Reply:
I think we need to handle this case. It'll be really confusing if someone updates the theme but nothing happens until they happen to rebuild (e.g. by rotating the phone).